README.md
![Build](../../workflows/Build/badge.svg)

# What is CosmOS?

CosmOS will be an [SSI](https://en.wikipedia.org/wiki/Single_system_image) operating system. The kernel of CosmOS is 64-bit, protected mode. Our strategy is:

- A single process space spanning multiple physical computers
- A single device space spanning multiple physical computers
- A single file system spanning multiple physical computers
- Automatic process migration between nodes
- Support for ARM and i386

## Screen Shot

![](doc/images/cosmos_screen_shot.png)

## Build and Run CosmOS

- [Building and Running CosmOS](doc/developer/build.md)

## Kernel Documentation

- [Source Tree](doc/developer/source_tree.md)
- [Devices](doc/developer/devices.md)
- [Device Status](doc/developer/device_status.md)
- [Working with Filesytems](doc/developer/working_with_filesystems.md)
- [Boot Disk Layout](doc/developer/boot_disk_layout.md)
- [VFS](doc/developer/vfs.md)
- [Using Block Devices](doc/developer/using_block_devices.md)
- [Block Devices](doc/developer/block_devices.md)
- [Memory Map](doc/developer/memory_map.md)
- [Collections](doc/developer/collections.md)
- [Assertions](doc/developer/assertions.md)
- [kprintf](doc/developer/kprintf_conversion_specifiers.md)
- [Interrupts](doc/developer/interrupts.md)
- [Console](doc/developer/kernel_terminal_console.md)
- [Keyboard](doc/developer/keyboard.md)
- [IO Buffers](doc/developer/io_buffers.md)
- [Strings](doc/developer/strings.md)
- [Syscalls](doc/developer/syscalls.md)
- File Systems
  - [Trivial File System (TFS)](doc/developer/filesystems/trivial_file_system.md)
  - [Cosmos File System (CFS) Disk Format](doc/developer/filesystems/cosmos_disk_format.md)

## Debugging

- [Debugging the kernel](doc/debugging/debug.md)

## Contributing

- [CosmOS coding standards](doc/contributing/coding_standards.md)
- [Operating System Development Resources](doc/contributing/osdev_resources.md)
- [cosmos-dev mailing list](https://www.freelists.org/list/cosmos-dev)

## License

- [License](LICENSE)


ChangeLog
2021-02-07
	* basic, non-useful but working system call interface (kmw)

2021-01-04
	* set up direct map and page directory (kmw)

2020-10-03
	* expand mapped heap space to 8 MB, begin work on slab allocator (kmw)

2020-9-24
	* added ata_channel_ready() to check if an ATA channel is ready for commands (kmw)

2020-09-22
	* begin DMA setup (kmw)

2020-09-20
	* determine logical sector size, moved ata device select into dedicated function (kmw)

2020-09-12
	* detect model, serial, and size of ATA devices (kmw)

2020-09-11
	* execute and read buffer from ATA IDENTIFY (PACKET) command (kmw)

2020-09-06
	* sleep_wait function (kmw)

2020-09-05
	* early ATA initialization, up to setting up and then turning off IRQs (kmw)

2020-08-30
	* in theory, bootloader should be able to handle kernels that go beyond the first track on disk (kmw)

2020-08-29
	* ascertain IDE control ports (kmw)

2020-08-23
	* string conversion class specifier for kprintf (kmw)

2020-08-22
	* added krealloc (kmw)
	* record PCI device info when scanning bus (kmw)
	* char conversion class specifier for kprintf (kmw)
	* added header_type field to pci_device_t (kmw)

2020-08-19
	* VGA console driver now interprets tab characters (kmw)
	* PCI bus scan (kmw)

2020-08-16
	* basic kprintf implementation--no format strings yet (kmw)
	* kprintf now supports unsigned integer-to-decimal conversions (kmw)
	* uint->hexadecimal conversion added to kprintf (kmw)
	* removed superseded debug_print_uint64 function now that kprintf is working (kmw)
	* memory blocks returned by kmalloc() are 8-byte aligned (kmw)

2020-08-15
	* console driver infrastructure and vga console driver (kmw)
	* removed old, superseded console subsystem (kmw)

2020-08-11
	* put check in vga text-mode driver to prevent writing past end of text-mode memory area (kmw)

2020-08-10
	* kernel video interface and vga text-mode driver (kmw)

2020-08-05
	* added raw keyboard driver (kmw)

2020-07-12:
	* update cursor location after writing to system console (kmw)

2020-07-11:
	* remap PIC, add skeleton IRQ handlers (kmw)
	* read system date and time from CMOS RTC (kmw)

2020-07-09:
	* added asm_[in, out]* functions (kmw)

2020-07-08:
	* fixed issue in bootloader (kmw)
	|> int 13h had been instructed to load beyond the 63 sector-per-track limit

2020-06-01:
	* kfree() will now join contiguous free blocks together (kmw)

2020-05-13:
	* added uitoa3() (kmw)
	|> this converts 64-bit unsigned integers (can be cast if shorter) to
	|> strings w/o needing dynamic memory allocation, which means it can be
	|> used before the mmu is initialized

2020-05-12:
	* added consWriteLine() (kmw)

user/bdos/bdos.h
//*****************************************************************
// This file is part of CosmOS                                    *
// Copyright (C) 2020-2021 Tom Everett                            *
// Released under the stated terms in the file LICENSE            *
// See the file "LICENSE" in the source distribution for details  *
// ****************************************************************

#ifndef _BDOS_H
#define _BDOS_H

// https://www.seasip.info/Cpm/bdosfunc.html
// https://www.seasip.info/Cpm/fcb.html

#include <cosmos_types.h>

struct bdos_fcb {
    uint8_t drive;
    uint8_t filename[8];
    uint8_t filetype[3];
    uint8_t ex;
    uint8_t s1;
    uint8_t s2;
    uint8_t rc;
    uint8_t allocation[16];
    uint8_t cr;
    uint8_t rn[3];
} __attribute__((packed));

void bdos_system_reset();
uint8_t bdos_console_input();
void bdos_console_output(uint8_t c);

void bdos_select_disk(uint8_t* name);
uint8_t bdos_open_file(struct bdos_fcb fcb);
uint8_t bdos_close_file(struct bdos_fcb fcb);
uint8_t bdos_delete_file(struct bdos_fcb fcb);
uint8_t bdos_make_file(struct bdos_fcb fcb);
uint8_t bdos_rename_file(struct bdos_fcb fcb);

/*
  0  X X X System Reset                 none           none
  1  X X X Console Input                none           A = char
  2  X X X Console Output               E = char       none
  3  X X - Auxiliary Input              none           A = char
  3  - - X Raw Console Input            none           A = char
  4  X X - Auxiliary Output             E = char       none
  4  - - X Raw Console Output           E = char       none
  5  X X X List Output                  E = char       none
  6  X X X Direct Console I/O           E = 0FFH/      A = char/
                                            0FEH/          status/
                                            0FDH/          none
                                            char
  7  X - - Get I/O Byte                 none           A = IOBYTE
  7  - X - Auxiliary Input Status       none           A = 00/0FFH
  8  X - - Set I/O Byte                 E = IOBYTE     none
  8  - X - Auxiliary Output Status      none           A = 00/0FFH
  9  X X X Print String                 DE = .String   none
 10  X X X Read Console Buffer          DE = .Buffer   Characters in
                                                           buffer
 11  X X X Get Console Status           none           A = 00/01
 12  X X X Return Version Number        none           HL = Version
 13  X X X Reset Disk System            none           see def
 14  X X X Select Disk                  E = Disk Number see def
 15  X X X Open File                    DE = .FCB      A = Dir Code
 16  X X X Close File                   DE = .FCB      A = Dir Code
 17  X X X Search for First             DE = .FCB      A = Dir Code
 18  X X X Search for Next              none           A = Dir Code
 19  X X X Delete File                  DE = .FCB      A = Dir Code
 20  X X X Read Sequential              DE = .FCB      A = Err Code
 21  X X X Write Sequential             DE = .FCB      A = Err Code
 22  X X X Make File                    DE = .FCB      A = Dir Code
 23  X X X Rename File                  DE = .FCB      A = Dir Code
 24  X X X Return Login Vector          none           HL= Login Vector
 25  X X X Return Current Disk          none           A = Cur Disk #
 26  X X X Set DMA Address              DE = .DMA      none
 27  X X X Get Addr (Alloc)             none           HL= .Alloc
 28  X X X Write Protect Disk           none           none
 29  X X X Get R/O Vector               none           HL= R/O Vector
 30  X X X Set File Attributes          DE = .FCB      A = Dir Code
 31  X X X Get Addr (DPB)               none           HL= .DPB
 32  X X X Set/Get User Code            E = 0FFH/      A = Curr User/
                                          user number      none
 33  X X X Read Random                  DE = .FCB      A = Err Code
 34  X X X Write Random                 DE = .FCB      A = Err Code
 35  X X X Compute File Size            DE = .FCB      r0, r1, r2
 36  X X X Set Random Record            DE = .FCB      r0, r1, r2
 37  X X X Reset Drive                  DE = Drive     A = Err Code
                                             Vector
 38  - X X Access Drive                 DE = Drive     none
                                             Vector
 39  - X X Free Drive                   DE = Drive     none
                                             Vector
 40  X X X Write Random with Zero Fill  DE = .FCB      A = Err Code
 41  - - X Test and Write Record        DE = .FCB      HL= Err Code
 41  - X - Test and Write Record        DE = .FCB      A = 0FFH
 42  - X X Lock Record                  DE = .FCB      HL = Err Code
 43  - X X Unlock Record                DE = .FCB      HL = Err Code
 44  - X X Set Multi-sector Count       E = # Sectors  A = Return Code
 45  - X X Set BDOS Error Mode          E = Err Mode   none
 46  - X X Get Disk Free Space          E = Drive #    Number of Free
                                                          Sectors
 47  - X X Chain to Program             E = Chain Flg  none
 48  - X X Flush Buffers                E = Purge Flg  A = Err Flag
 49  - X - Get/Set System Control       DE = .SCB PB   A = Returned Byte
                         Block                         HL= Returned Word
 50  - X - Direct Bios Calls            DE = .BIOS PB  BIOS Return
 59  - X - Load Overlay                 DE = .FCB      A = Err Code
 60  - X - Call Resident System         DE = .RSX PB   A = Err Code
                    Extension
 98  - X - Free Blocks                  none           none
 99  - X - Truncate File                DE = .FCB      A = Dir Code
100  - X X Set Directory Label          DE = .FCB      A = Dir Code
101  - X X Return Directory Label Data  E = Drive      A = Label Data
102  - X X Read File Date Stamps        DE = .FCB      A = Dir Code
               and Password Mode
103  - X X Write File XFCB              DE = .FCB      A = Dir Code
104  - X X Set Date and Time            DE = .DAT      none
105  - X X Get Date and Time            DE = .DAT      Date and Time
                                                       A = seconds
106  - X X Set Default Password         DE = .Passwrd  none
107  - X X Return Serial Number         DE = .Serial # Serial Number
                                             field
108  - X - Get/Set Program Return Code  DE = 0FFFFH/   HL= Ret Code/
                                             Code          none
109  - X - Get/Set Console Mode         DE = 0FFFFH/   HL= Console Mode/
                                             Mode          none
110  - X - Get/Set Output Delimiter     DE = 0FFFFH/   A = Output Delim/
                                        E = Delimiter      none
111  - X - Print Block                  DE = .CCB      none
112  - X - List Block                   DE = .CCB      none
128  - - X Absolute Memory Request      DE = .MD       A = Err Code
129  - - X Relocatable Memory Request   DE = .MD       A = Err Code
130  - - X Memory Free                  DE = .MD       none
131  - - X Poll                         E = Device     none
132  - - X Flag Wait                    E = Flag       A = Err Code
133  - - X Flag Set                     E = Flag       A = Err Code
134  - - X Make Queue                   DE = .QCB      none
135  - - X Open Queue                   DE = .UQCB     A = Err Code
136  - - X Delete Queue                 DE = .QCB      A = Err Code
137  - - X Read Queue                   DE = .UQCB     none
138  - - X Conditional Read Queue       DE = .UQCB     A = Err Code
139  - - X Write Queue                  DE = .UQCB     none
140  - - X Conditional Write Queue      DE = .UQCB     A = Err Code
141  - - X Delay                        DE = # ticks   none
142  - - X Dispatch                     none           none
143  - - X Terminate Process            E = Term. Code none
144  - - X Create Process               DE = .PD       none
145  - - X Set Priority                 E = Priority   none
146  - - X Attach Console               none           none
147  - - X Detach Console               none           none
148  - - X Set Console                  E = Console    none
149  - - X Assign Console               DE = .APB      A = Err Code
150  - - X Send CLI Command             DE = .CLICMD   none
151  - - X Call Resident Sys Proc       DE = .CPB      HL= result
152  - X X Parse Filename               DE = .PFCB     see def
153  - - X Get Console Number           none           A = Console #
154  - - X System Data Address          none           HL= Sys Data Addr
155  - - X Get Date and Time            DE = .TOD      date and time
156  - - X Return Process Descriptor    none           HL= PD Addr
157  - - X Abort Specified Process      DE = .ABT PB   A = Err Code
158  - - X Attach List                  none           none
159  - - X Detach List                  none           none
160  - - X Set List                     E = List #     none
161  - - X Conditional Attach List      none           A = Err Code
162  - - X Conditional Attach Console   none           A = Err Code
163  - - X MPM Version Number           none           HL= Version #
164  - - X Get List Number              none           A = List #

*/
#endif